cosmosapi/get_partition_key_ranges.go

```diff
@@ -37,15 +55,139 @@ type PartitionKeyRange struct {
+	options.MaxItemCount = -1
+	options.Continuation = ""
+	p := c.NewPartitionKeyRangesPaginator(databaseName, collectionName, options)
+	var pkranges GetPartitionKeyRangesResponse
+	for p.Next() {
+		newPk, err := p.CurrentPage(ctx)
+		if err != nil {
+			return pkranges, err
+		}
+		newPk.PartitionKeyRanges = append(pkranges.PartitionKeyRanges, newPk.PartitionKeyRanges...)
+		newPk.RequestCharge += pkranges.RequestCharge
+		pkranges = newPk
+	}
+	return pkranges, nil
+}
+
+// NewPartitionKeyRangesPaginator returns a paginator for ListObjectsV2. Use the
+// Next method to get the next page, and CurrentPage to get the current response
+// page from the paginator. Next will return false if there are no more pages,
+// or an error was encountered.
+//
+// Note: This operation can generate multiple requests to a service.
+//
+//   // Example iterating over pages.
+//   p := client.NewPartitionKeyRangesPaginator(input)
+//
+//   for p.Next() {
+//       err, page := p.CurrentPage(context.TODO())
+//       if err != nil {
+//         return err
+//       }
+//   }
+//
+func (c *Client) NewPartitionKeyRangesPaginator(databaseName, collectionName string, options *GetPartitionKeyRangesOptions) *PartitionKeyRangesPaginator {
+	var opts GetPartitionKeyRangesOptions
+	if options != nil {
+		opts = *options
+	}
+	return &PartitionKeyRangesPaginator{
+		databaseName:   databaseName,
+		collectionName: collectionName,
+		options:        opts,
+		client:         c,
+	}
+}
+
+// PartitionKeyRangesPaginator is a paginator over the "Get Partition key
+// ranges" API endpoint. This paginator is not threadsafe.
+type PartitionKeyRangesPaginator struct {
+	shouldFetchPage bool
+	hasPage         bool
+
+	err         error
+	currentPage GetPartitionKeyRangesResponse
+
+	client         *Client
+	databaseName   string
+	collectionName string
+	options        GetPartitionKeyRangesOptions
+}
+
+// CurrentPage returns the current page of partition key ranges. Panics if
+// Next() has not yet been called.
+func (p *PartitionKeyRangesPaginator) CurrentPage(ctx context.Context) (GetPartitionKeyRangesResponse, error) {
+	if !p.shouldFetchPage && !p.hasPage {
+		panic("PartitionKeyRangesPaginator: Must call Next before CurrentPage")
+	}
+	if p.shouldFetchPage { // includes retries if the previous call errored out
+		p.currentPage, p.err = p.client.GetPartitionKeyRanges(ctx, p.databaseName, p.collectionName, &p.options)
+		if p.err == nil {
+			p.shouldFetchPage = false
+			p.hasPage = true
+			p.options.Continuation = p.currentPage.Continuation
+		}
+	}
+	return p.currentPage, p.err
+}
+
+// Next returns true if there are more pages to be read, and false if the
+// previous CurrentPage call returned an error, or if there are no more pages to
+// be read.
+func (p *PartitionKeyRangesPaginator) Next() bool {
+	if p.err != nil {
+		return false
+	}
+	if !p.hasPage {
```

Comment:
Just a control question: this scope/code branch is executed at max _one time_, right. After the first page has been fetched, `hasPage` remains true. Correct?

Reply:
You actually did answer the question already in your comment above. I am good then.